Autor:  
Manuel Eugenio Morocho Cayamcela, PhD

## Procesamiento y Transformación de Datos para Análisis de Crédito

### **Introducción:**

Esta lectura cubrirá el preprocesamiento y transformación de datos para el análisis de crédito. Utilizaremos un conjunto de datos de solicitudes de crédito con las siguientes columnas:

1. ID: Identificador único para cada registro.
2. Customer_ID: Identificador único para cada cliente.
3. Month: Mes en el que se registró la información.
4. Name: Nombre del cliente.
5. Age: Edad del cliente.
6. SSN: Número de Seguro Social del cliente.
7. Occupation: Ocupación del cliente.
8. Annual_Income: Ingreso anual del cliente.
9. Monthly_Inhand_Salary: Salario mensual disponible del cliente.
10. Num_Bank_Accounts: Número de cuentas bancarias del cliente.
11. Num_Credit_Card: Número de tarjetas de crédito del cliente.
12. Interest_Rate: Tasa de interés aplicada al cliente.
13. Num_of_Loan: Número de préstamos del cliente.
14. Type_of_Loan: Tipo de préstamo(s) del cliente.
15. Delay_from_due_date: Retraso desde la fecha de vencimiento del pago.
16. Num_of_Delayed_Payment: Número de pagos retrasados del cliente.
17. Changed_Credit_Limit: Límite de crédito modificado del cliente.
18. Num_Credit_Inquiries: Número de consultas de crédito realizadas por el cliente.
19. Credit_Mix: Mezcla de crédito del cliente.
20. Outstanding_Debt: Deuda pendiente del cliente.
21. Credit_Utilization_Ratio: Relación de utilización de crédito del cliente.
22. Credit_History_Age: Antigüedad del historial de crédito del cliente.
23. Payment_of_Min_Amount: Pago del monto mínimo realizado por el cliente.
24. Total_EMI_per_month: Total de cuotas mensuales del cliente.
25. Amount_invested_monthly: Cantidad invertida mensualmente por el cliente.
26. Payment_Behaviour: Comportamiento de pago del cliente.
27. Monthly_Balance: Saldo mensual del cliente.
28. Credit_Score: Puntuación de crédito del cliente.

### **Contenido:**

#### Importamos las librerías necesarias

In [46]:
import pandas as pd
import numpy as np

#### Carga de Datos

In [53]:
# Cargamos y mostramos los primeros registros de la base de datos credit_score.csv
data = pd.read_csv('/Users/eugenio/Documents/Notebooks_ArtificialIntelligence/Datasets/credit_score.csv')
data.head(10)

/var/folders/8j/f0xtldm91_1c2bpdr69fxgzw0000gn/T/ipykernel_83369/2464886221.py:2: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('/Users/eugenio/Documents/Notebooks_ArtificialIntelligence/Datasets/credit_score.csv')


,ID,Customer_ID,Month,Name,Age,SSN,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
0,0x1602,CUS_0xd40,January,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,1824.843333,3,...,_,809.98,26.822620,22 Years and 1 Months,No,49.574949,80.41529543900253,High_spent_Small_value_payments,312.49408867943663,Good
1,0x1603,CUS_0xd40,February,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,31.944960,NaN,No,49.574949,118.28022162236736,Low_spent_Large_value_payments,284.62916249607184,Good
2,0x1604,CUS_0xd40,March,Aaron Maashoh,-500,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,28.609352,22 Years and 3 Months,No,49.574949,81.699521264648,Low_spent_Medium_value_payments,331.2098628537912,Good
3,0x1605,CUS_0xd40,April,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,31.377862,22 Years and 4 Months,No,49.574949,199.4580743910713,Low_spent_Small_value_payments,223.45130972736786,Good
4,0x1606,CUS_0xd40,May,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,1824.843333,3,...,Good,809.98,24.797347,22 Years and 5 Months,No,49.574949,41.420153086217326,High_spent_Medium_value_payments,341.48923103222177,Good
5,0x1607,CUS_0xd40,June,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,NaN,3,...,Good,809.98,27.262259,22 Years and 6 Months,No,49.574949,62.430172331195294,!@9#%8,340.4792117872438,Good
6,0x1608,CUS_0xd40,July,Aaron Maashoh,23,821-00-0265,Scientist,19114.12,1824.843333,3,...,Good,809.98,22.537593,22 Years and 7 Months,No,49.574949,178.3440674122349,Low_spent_Small_value_payments,244.5653167062043,Good
7,0x1609,CUS_0xd40,August,NaN,23,#F%$D@*&8,Scientist,19114.12,1824.843333,3,...,Good,809.98,23.933795,NaN,No,49.574949,24.785216509052056,High_spent_Medium_value_payments,358.12416760938714,Standard
8,0x160e,CUS_0x21b1,January,Rick Rothackerj,28_,004-07-5839,_______,34847.84,3037.986667,2,...,Good,605.03,24.464031,26 Years and 7 Months,No,18.816215,104.291825168246,Low_spent_Small_value_payments,470.69062692529184,Standard
9,0x160f,CUS_0x21b1,February,Rick Rothackerj,28,004-07-5839,Teacher,34847.84,3037.986667,2,...,Good,605.03,38.550848,26 Years and 8 Months,No,18.816215,40.39123782853101,High_spent_Large_value_payments,484.5912142650067,Good


#### Exploración de Tipos de Datos


Para entender qué tipo de datos tenemos, podemos usar el método `info()`.

In [48]:
# Mostramos la información de la base de datos 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      90015 non-null   object 
 4   Age                       100000 non-null  object 
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  object 
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

**Observaciones:**
1. Hay valores faltantes presentes en el conjunto de datos.
2. El conjunto de datos tiene valores `object` y `float64`.

In [51]:
# Mostramos la cantidad de registros y columnas de la base de datos
data.describe(exclude=np.number).T

,count,unique,top,freq
ID,100000,100000,0x1602,1
Customer_ID,100000,12500,CUS_0xd40,8
Month,100000,8,January,12500
Name,90015,10139,Langep,44
Age,100000,1788,38,2833
SSN,100000,12501,#F%$D@*&8,5572
Occupation,100000,16,_______,7062
Annual_Income,100000,18940,36585.12,16
Num_of_Loan,100000,434,3,14386
Type_of_Loan,88592,6260,Not Specified,1408


**Observaciones:**

1. `Customer_ID` tiene 12500 valores únicos. Esto significa que tenemos datos de 12500 clientes.
2. `Month` solo tiene 8 valores únicos. Debemos analizar más a fondo qué meses están presentes.
3. `Age` tiene 1788 valores únicos. Esto parece extraño ya que el rango general de edad es de 0-100.
4. `SSN` tiene 12501 valores únicos, mientras que `Customer_ID` solo tiene 12500 valores únicos. Existe la posibilidad de que se haya ingresado un valor incorrecto de SSN para uno de los clientes, ya que la misma persona no puede tener múltiples SSN.

#### Identificación de Valores Faltantes

In [54]:
# Identificar valores faltantes
missing_values = data.isnull().sum()
missing_values

ID                              0
Customer_ID                     0
Month                           0
Name                         9985
Age                             0
SSN                             0
Occupation                      0
Annual_Income                   0
Monthly_Inhand_Salary       15002
Num_Bank_Accounts               0
Num_Credit_Card                 0
Interest_Rate                   0
Num_of_Loan                     0
Type_of_Loan                11408
Delay_from_due_date             0
Num_of_Delayed_Payment       7002
Changed_Credit_Limit            0
Num_Credit_Inquiries         1965
Credit_Mix                      0
Outstanding_Debt                0
Credit_Utilization_Ratio        0
Credit_History_Age           9030
Payment_of_Min_Amount           0
Total_EMI_per_month             0
Amount_invested_monthly      4479
Payment_Behaviour               0
Monthly_Balance              1200
Credit_Score                    0
dtype: int64

In [55]:
# Mostramos el porcentaje de valores faltantes
data.isnull().mean()*100

ID                           0.000
Customer_ID                  0.000
Month                        0.000
Name                         9.985
Age                          0.000
SSN                          0.000
Occupation                   0.000
Annual_Income                0.000
Monthly_Inhand_Salary       15.002
Num_Bank_Accounts            0.000
Num_Credit_Card              0.000
Interest_Rate                0.000
Num_of_Loan                  0.000
Type_of_Loan                11.408
Delay_from_due_date          0.000
Num_of_Delayed_Payment       7.002
Changed_Credit_Limit         0.000
Num_Credit_Inquiries         1.965
Credit_Mix                   0.000
Outstanding_Debt             0.000
Credit_Utilization_Ratio     0.000
Credit_History_Age           9.030
Payment_of_Min_Amount        0.000
Total_EMI_per_month          0.000
Amount_invested_monthly      4.479
Payment_Behaviour            0.000
Monthly_Balance              1.200
Credit_Score                 0.000
dtype: float64

#### Eliminación de columnas que no aportan información útil

La decisión de qué columnas eliminar depende en gran medida del contexto y del objetivo del análisis. Sin embargo, aquí hay algunas pautas generales que podemos considerar:

**Irrelevancia para el modelado**: Las columnas que contienen un valor único para cada fila (como un ID) generalmente no aportan información útil para un modelo de data mining y pueden ser eliminadas.

Vamos a usar la función `drop` de pandas para eliminar las columnas con valores únicos. 
`data.drop(...)`: Esta es la función de pandas que se utiliza para eliminar columnas o filas de un DataFrame.

Por ejemplo, la columna `'Name'` contiene los nombres de los individuos que han solicitado créditos. En general, los nombres de las personas no proporcionan información útil para un modelo de aprendizaje automático, ya que son únicos para cada individuo y no tienen una relación directa con otras variables.

Además, codificar columnas con valores únicos, podría resultar en un gran número de características únicas, lo que podría hacer el modelo sea más complejo y difícil de entrenar.

Por lo tanto, en la mayoría de los casos, es seguro eliminar la columna 'Name' del conjunto de datos. Sin embargo, hay excepciones. Por ejemplo, si los nombres tienen algún tipo de información codificada (como un prefijo que indica el género o el estatus social), podemos extraer esa información antes de eliminar la columna. Pero en la mayoría de los casos, especialmente si estás tratando con nombres reales de personas, es mejor eliminar la columna.

Las columnas `ID`, `SSN` y `Customer_ID` se eliminan generalmente por las siguientes razones:

**Irrelevancia para el modelado**: Estas columnas suelen ser identificadores únicos para cada registro o cliente. No contienen información útil para el modelado predictivo, ya que no tienen ninguna relación con las otras variables.

**Evitar el sobreajuste**: Si se incluyen en el modelo, pueden causar sobreajuste. El modelo podría aprender a hacer predicciones basándose en el ID del cliente, lo cual no es útil ni generalizable.

**Privacidad de los datos**: En algunos casos, los identificadores de los clientes se eliminan para proteger la privacidad de los clientes, especialmente cuando los datos se comparten o se publican.

In [56]:
# Eliminamos columnas que no aportan información relevante para el análisis
data = data.drop(['ID', 'SSN', 'Customer_ID', 'Name', 'Month'], axis=1) # Se elimina month ya que no aporta información relevante para el análisis
data.head()

,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,Delay_from_due_date,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
0,23,Scientist,19114.12,1824.843333,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,...,_,809.98,26.822620,22 Years and 1 Months,No,49.574949,80.41529543900253,High_spent_Small_value_payments,312.49408867943663,Good
1,23,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",-1,...,Good,809.98,31.944960,NaN,No,49.574949,118.28022162236736,Low_spent_Large_value_payments,284.62916249607184,Good
2,-500,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,...,Good,809.98,28.609352,22 Years and 3 Months,No,49.574949,81.699521264648,Low_spent_Medium_value_payments,331.2098628537912,Good
3,23,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",5,...,Good,809.98,31.377862,22 Years and 4 Months,No,49.574949,199.4580743910713,Low_spent_Small_value_payments,223.45130972736786,Good
4,23,Scientist,19114.12,1824.843333,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",6,...,Good,809.98,24.797347,22 Years and 5 Months,No,49.574949,41.420153086217326,High_spent_Medium_value_payments,341.48923103222177,Good


#### Tratamiento de la columna `Age`

In [18]:
# Mostramos los valores únicos de la columna 'Age'
print(list(data['Age'].unique()))

['23', '-500', '28_', '28', '34', '54', '55', '21', '31', '33', '34_', '7580', '30', '30_', '24', '24_', '44', '45', '40', '41', '32', '33_', '35', '35_', '36', '39', '37', '181', '20', '46', '26', '42', '19', '31_', '48', '995', '40_', '37_', '38', '54_', '5079', '43', '21_', '22', '6409', '16', '7080', '18', '849', '3885', '20_', '15', '27', '43_', '38_', '25', '3052', '14', '5342', '17', '18_', '4431', '2657', '2111_', '46_', '47', '1032', '16_', '19_', '47_', '456', '5717', '53_', '53', '56', '25_', '27_', '55_', '3169', '1191', '29', '42_', '7140', '48_', '49', '49_', '8233', '6955', '2534', '3115', '7657', '51', '50', '5112', '50_', '32_', '6452', '2744', '22_', '17_', '1439', '29_', '5795', '4872', '1772', '15_', '1383', '5657', '52', '1934', '51_', '8352', '3734', '26_', '2056', '39_', '7508', '2339', '14_', '8406', '1424', '36_', '6953', '5626', '4471', '548', '44_', '769', '5490', '525', '4202', '3665', '7670_', '4670', '3616', '6922', '6619', '1808', '7992', '45_', '223', '1

**Observación:**
1. Algunos valores de `Age` tienen un guion bajo, que puede representar un error de tipeo (inconsistencia).
2. Se encuentran valores de edad negativos, lo que no es posible.
3. Aunque es posible que un cliente tenga más de 100 años, necesitamos establecer un rango.

In [19]:
# Substituimos los valores '_' por '' y los convertimos a tipo flotante ('float')
data['Age'] = data['Age'].replace('_', '', regex=True).astype('float') # regex=True para que reemplace todas las ocurrencias. regex means regular expression.

# Definimos un rango para los valores de la columna 'Age' de 1:100 y sustituimos los restantes con 'nan'
data.loc[data.Age < 1, 'Age'] = np.nan
data.loc[data.Age > 100, 'Age'] = np.nan

# Reemplazamos los valores 'nan' de la columna 'Age' con la media de la misma
data['Age'] = data['Age'].fillna(data['Age'].mean())

# Mostramos los valores únicos de la columna 'Age'
print(list(data['Age'].unique()))

# Mostramos el tipo de la columna 'Age'
print(data['Age'].dtype)

[23.0, 33.32327408870238, 28.0, 34.0, 54.0, 55.0, 21.0, 31.0, 33.0, 30.0, 24.0, 44.0, 45.0, 40.0, 41.0, 32.0, 35.0, 36.0, 39.0, 37.0, 20.0, 46.0, 26.0, 42.0, 19.0, 48.0, 38.0, 43.0, 22.0, 16.0, 18.0, 15.0, 27.0, 25.0, 14.0, 17.0, 47.0, 53.0, 56.0, 29.0, 49.0, 51.0, 50.0, 52.0, 99.0, 100.0, 95.0]
float64


#### Tratamiento de la columna `Occupation`

In [20]:
# Mostramos los valores únicos de la columna 'Occupation'
print(list(data['Occupation'].unique()))

['Scientist', '_______', 'Teacher', 'Engineer', 'Entrepreneur', 'Developer', 'Lawyer', 'Media_Manager', 'Doctor', 'Journalist', 'Manager', 'Accountant', 'Musician', 'Mechanic', 'Writer', 'Architect']


**Observación:**
1. Existe un valor '_______' que se debe eliminar.

In [57]:
# Substituimos el valor '_______' por 'nan' y los convertimos a tipo flotante ('float')
data['Occupation'] = data['Occupation'].replace('_______', np.nan, regex=True)

# Mostramos el número e valores 'nan' en la columna 'Occupation'
print(data['Occupation'].isna().sum())

7062


**Observación:**
1. Tenemos un número considerable de registros `NaN`. Tenemos dos opciones: i) Reemplazas los valores `NaN` con la moda, no perderás ningún registro. Pero en este contexto, estaríamos imponiendo una profesión a clientes que posiblemente no llenaron ese campo. ii) La segunda opción es eliminar los registros NaN, perderás esos registros, pero tus datos serán más precisos.

- Opción 1: Reemplazar los datos con la moda:



In [ ]:
# Calculamos la moda
# mode = data['Occupation'].mode()[0]
# mode

'Lawyer'

In [23]:
# Reemplazamos los valores NaN con la moda
# data['Occupation'].fillna(mode, inplace=True)

- Opción 2: Eliminar los registros NaN:

In [59]:
# Eliminamos los registros con valores NaN en la columna 'Occupation'
data.dropna(subset=['Occupation'], inplace=True)

**Eliminar columnas con muchos valores faltantes**: Si una columna tiene una gran cantidad de valores faltantes (~>70%), podría ser más útil eliminarla que intentar imputar los valores faltantes.

En este ejemplo, no existen columnas con más del 70% de valores faltantes. De existir, podemos eliminarla con el siguiente código:

In [25]:
# Eliminar columnas del DataFrame 'data' con más del 70% de valores faltantes
threshold = int(len(data) * 0.7) # Calcular el umbral de valores faltantes
data = data.dropna(thresh=threshold, axis=1) # Eliminar columnas con más del 70% de valores faltantes

**Reemplazar valores faltantes en columnas con pocos valores faltantes**: Los valores faltantes pueden ser identificados con el método `isnull()`. Para manejarlos, podemos usar la imputación (rellenar los valores faltantes con la media, mediana, o moda) o simplemente eliminar las filas con valores faltantes.

Para rellenar con la media los valores faltantes de las columnas numéricas con menos del 10% de valores faltantes, puedes usar la biblioteca pandas. Aquí te dejo un ejemplo de cómo hacerlo:



In [26]:
# DataFrame
for col in data.columns:
    if data[col].isnull().sum() / len(data) < 0.1:
        if data[col].dtype == 'float64' or data[col].dtype == 'int64':
            data[col].fillna(data[col].mean(), inplace=True)

# Mostramos el porcentaje de valores faltantes
data.isnull().mean()*100

/var/folders/8j/f0xtldm91_1c2bpdr69fxgzw0000gn/T/ipykernel_83369/2893222039.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[col].fillna(data[col].mean(), inplace=True)


Age                          0.000000
Occupation                   0.000000
Annual_Income                0.000000
Monthly_Inhand_Salary       15.001399
Num_Bank_Accounts            0.000000
Num_Credit_Card              0.000000
Interest_Rate                0.000000
Num_of_Loan                  0.000000
Type_of_Loan                11.344122
Delay_from_due_date          0.000000
Num_of_Delayed_Payment       6.985302
Changed_Credit_Limit         0.000000
Num_Credit_Inquiries         0.000000
Credit_Mix                   0.000000
Outstanding_Debt             0.000000
Credit_Utilization_Ratio     0.000000
Credit_History_Age           9.044740
Payment_of_Min_Amount        0.000000
Total_EMI_per_month          0.000000
Amount_invested_monthly      4.476102
Payment_Behaviour            0.000000
Monthly_Balance              1.210484
Credit_Score                 0.000000
dtype: float64



Este código recorre cada columna en el DataFrame. Si la columna tiene menos del 10% de valores faltantes y es numérica (es decir, su tipo de datos es 'float64' o 'int64'), entonces rellena los valores faltantes con la media de la columna.

La imputación de la media no es aplicable a las columnas de tipo string (categóricas). Para las columnas categóricas, puedes considerar diferentes métodos de imputación, como el más frecuente (moda), una constante o una cadena de texto específica como 'Desconocido'. Aquí te dejo un ejemplo de cómo hacerlo con pandas:

In [28]:
# Rellenar con la moda
categorical_columns = data.select_dtypes(include=[object]).columns # seleccionar columnas categóricas
data[categorical_columns] = data[categorical_columns].fillna(data[categorical_columns].mode().iloc[0]) # rellenar con la moda. iloc[0] se utiliza para seleccionar la primera fila del DataFrame resultante de mode(), que contiene el valor más frecuente para cada columna categórica. Esto asegura que se rellenen los valores faltantes con la moda correcta para cada columna.

**Columnas con alta correlación**: Si tienes dos columnas que están altamente correlacionadas, podrías considerar eliminar una de ellas. La razón es que ambas contienen información similar y tener ambas en el análisis podría introducir multicolinealidad.



Recuerda que estas son solo pautas generales y la decisión final de qué columnas eliminar debe basarse en el conocimiento del dominio y el objetivo de tu análisis.


**Codificación de Variables Categóricas**: Las variables categóricas como 'Occupation' y 'Type_of_Loan' deben ser codificadas antes de ser usadas en un modelo. Podrías usar codificación one-hot o codificación de etiquetas dependiendo de la naturaleza de los datos.

In [60]:
# Codificación one-hot
data_encoded = pd.get_dummies(data, columns=['Occupation', 'Type_of_Loan'])

In [63]:
data_encoded

,Age,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Delay_from_due_date,Num_of_Delayed_Payment,Changed_Credit_Limit,...,"Type_of_Loan_Student Loan, Student Loan, and Student Loan","Type_of_Loan_Student Loan, and Auto Loan","Type_of_Loan_Student Loan, and Credit-Builder Loan","Type_of_Loan_Student Loan, and Debt Consolidation Loan","Type_of_Loan_Student Loan, and Home Equity Loan","Type_of_Loan_Student Loan, and Mortgage Loan","Type_of_Loan_Student Loan, and Not Specified","Type_of_Loan_Student Loan, and Payday Loan","Type_of_Loan_Student Loan, and Personal Loan","Type_of_Loan_Student Loan, and Student Loan"
0,23,19114.12,1824.843333,3,4,3,4,3,7,11.27,...,False,False,False,False,False,False,False,False,False,False
1,23,19114.12,NaN,3,4,3,4,-1,NaN,11.27,...,False,False,False,False,False,False,False,False,False,False
2,-500,19114.12,NaN,3,4,3,4,3,7,_,...,False,False,False,False,False,False,False,False,False,False
3,23,19114.12,NaN,3,4,3,4,5,4,6.27,...,False,False,False,False,False,False,False,False,False,False
4,23,19114.12,1824.843333,3,4,3,4,6,NaN,11.27,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99995,25,39628.99,3359.415833,4,6,7,2,23,7,11.5,...,False,False,False,False,False,False,False,False,False,False
99996,25,39628.99,3359.415833,4,6,7,2,18,7,11.5,...,False,False,False,False,False,False,False,False,False,False
99997,25,39628.99,3359.415833,4,6,5729,2,27,6,11.5,...,False,False,False,False,False,False,False,False,False,False
99998,25,39628.99,3359.415833,4,6,7,2,20,NaN,11.5,...,False,False,False,False,False,False,False,False,False,False


**Escalado de Características**: Las características como 'Annual_Income', 'Monthly_Inhand_Salary', 'Outstanding_Debt' podrían tener diferentes rangos y podrían necesitar ser escaladas.

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
data_scaled = scaler.fit_transform(data[['Annual_Income', 'Monthly_Inhand_Salary', 'Outstanding_Debt']])

**Selección de Características**: Podrías usar métodos como la correlación de Pearson para identificar las variables más relevantes para tu variable objetivo.

In [ ]:
correlation = data.corr()
correlation['Credit_Score'].sort_values(ascending=False)


#### Identificación y Manejo de Outliers

Los outliers pueden ser identificados utilizando métodos estadísticos, como el método del rango intercuartil (IQR). Una vez identificados, pueden ser eliminados o ajustados.



In [ ]:
# Identificar outliers con el método IQR
Q1 = data.quantile(0.25)
Q3 = data.quantile(0.75)
IQR = Q3 - Q1

# Eliminar outliers
data = data[~((data < (Q1 - 1.5 * IQR)) | (data > (Q3 + 1.5 * IQR))).any(axis=1)]

#### Identificación y Manejo de Datos Inconsistentes

Los datos inconsistentes pueden ser identificados revisando los valores únicos de cada característica. Una vez identificados, pueden ser corregidos o eliminados.



In [33]:
# Identificar datos inconsistentes
for column in data.columns:
    print(data[column].unique())

[ 23.          33.32327409  28.          34.          54.
  55.          21.          31.          33.          30.
  24.          44.          45.          40.          41.
  32.          35.          36.          39.          37.
  20.          46.          26.          42.          19.
  48.          38.          43.          22.          16.
  18.          15.          27.          25.          14.
  17.          47.          53.          56.          29.
  49.          51.          50.          52.          99.
 100.          95.        ]
['Scientist' 'Teacher' 'Engineer' 'Entrepreneur' 'Developer' 'Lawyer'
 'Media_Manager' 'Doctor' 'Journalist' 'Manager' 'Accountant' 'Musician'
 'Mechanic' 'Writer' 'Architect']
['19114.12' '34847.84' '34847.84_' ... '20002.88' '39628.99' '39628.99_']
[1824.84333333           nan 3037.98666667 ... 3097.00833333 1929.90666667
 3359.41583333]
[   3    2    1    7    4    0    8    5    6    9   10 1414 1231   67
 1488   91  528 1647 1696 1338  649  

#### Balanceo de Datos

El balanceo de datos puede ser necesario si nuestra variable objetivo está desbalanceada. Podemos usar técnicas como el oversampling o el undersampling.



In [ ]:
from imblearn.over_sampling import SMOTE

smote = SMOTE()
X_res, y_res = smote.fit_resample(data.drop('target', axis=1), data['target'])

#### Discretización

La discretización convierte las variables continuas en variables discretas.

In [ ]:
data['age'] = pd.cut(data['age'], bins=3, labels=['young', 'middle-aged', 'old'])



Finalmente, podemos usar una regresión lineal (o cualquier otro modelo de aprendizaje automático) para predecir si a un cliente se le otorgará crédito o no.



In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_res, y_res)



Este tutorial es solo una introducción a las técnicas de procesamiento y transformación de datos. Hay muchas otras técnicas y métodos disponibles, y la elección de las técnicas a utilizar dependerá de la naturaleza de tus datos y del problema que estés tratando de resolver.

## **No time?** Try this...

In [65]:
import pandas as pd
data = pd.read_csv('/Users/eugenio/Documents/Notebooks_ArtificialIntelligence/Datasets/credit_score.csv')
data = data.drop(['ID', 'Customer_ID', 'Month', 'Name', 'SSN',], axis=1)
data.head()

/var/folders/8j/f0xtldm91_1c2bpdr69fxgzw0000gn/T/ipykernel_83369/3576933381.py:2: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('/Users/eugenio/Documents/Notebooks_ArtificialIntelligence/Datasets/credit_score.csv')


,Age,Occupation,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,Delay_from_due_date,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,Credit_Score
0,23,Scientist,19114.12,1824.843333,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,...,_,809.98,26.822620,22 Years and 1 Months,No,49.574949,80.41529543900253,High_spent_Small_value_payments,312.49408867943663,Good
1,23,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",-1,...,Good,809.98,31.944960,NaN,No,49.574949,118.28022162236736,Low_spent_Large_value_payments,284.62916249607184,Good
2,-500,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",3,...,Good,809.98,28.609352,22 Years and 3 Months,No,49.574949,81.699521264648,Low_spent_Medium_value_payments,331.2098628537912,Good
3,23,Scientist,19114.12,NaN,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",5,...,Good,809.98,31.377862,22 Years and 4 Months,No,49.574949,199.4580743910713,Low_spent_Small_value_payments,223.45130972736786,Good
4,23,Scientist,19114.12,1824.843333,3,4,3,4,"Auto Loan, Credit-Builder Loan, Personal Loan,...",6,...,Good,809.98,24.797347,22 Years and 5 Months,No,49.574949,41.420153086217326,High_spent_Medium_value_payments,341.48923103222177,Good


In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 23 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Age                       100000 non-null  object 
 1   Occupation                100000 non-null  object 
 2   Annual_Income             100000 non-null  object 
 3   Monthly_Inhand_Salary     84998 non-null   float64
 4   Num_Bank_Accounts         100000 non-null  int64  
 5   Num_Credit_Card           100000 non-null  int64  
 6   Interest_Rate             100000 non-null  int64  
 7   Num_of_Loan               100000 non-null  object 
 8   Type_of_Loan              88592 non-null   object 
 9   Delay_from_due_date       100000 non-null  int64  
 10  Num_of_Delayed_Payment    92998 non-null   object 
 11  Changed_Credit_Limit      100000 non-null  object 
 12  Num_Credit_Inquiries      98035 non-null   float64
 13  Credit_Mix                100000 non-null  ob

In [37]:
# Identify categorical columns
cat_cols = data.select_dtypes(include=["object"]).columns
print(cat_cols)

Index(['Age', 'Occupation', 'Annual_Income', 'Num_of_Loan', 'Type_of_Loan',
       'Num_of_Delayed_Payment', 'Changed_Credit_Limit', 'Credit_Mix',
       'Outstanding_Debt', 'Credit_History_Age', 'Payment_of_Min_Amount',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')


In [67]:
# clean categorical columns
for col in cat_cols:
    data[col] = data[col].str.strip().str.lower() # .strip removes leading and trailing whitespace, and .lower() converts to lowercase
    data[col] = data[col].replace(["?", "na", "none", "null", "","_","_______"], pd.NA)
    mode = data[col].mode()[0] 
    data[col].fillna(mode, inplace=True)

/var/folders/8j/f0xtldm91_1c2bpdr69fxgzw0000gn/T/ipykernel_83369/2427841289.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[col].fillna(mode, inplace=True)


In [39]:
data.shape

(100000, 23)

In [68]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
for col in cat_cols:
    data[col] = data[col].astype(str) # avoid errors with NA
    data[col] = le.fit_transform(data[col])

In [41]:
data.shape

(100000, 23)

In [69]:
# Apply coerce to all numeric columns
for col in data.select_dtypes(include=["number"]).columns:
    data[col] = pd.to_numeric(data[col], errors='coerce') # errors='coerce' will convert invalid parsing to NaN

In [70]:
# Mean imputation
for col in data.select_dtypes(include=['float64', 'int64']).columns:
    data[col].fillna(data[col].mean(), inplace=True)

/var/folders/8j/f0xtldm91_1c2bpdr69fxgzw0000gn/T/ipykernel_83369/3347571600.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[col].fillna(data[col].mean(), inplace=True)


In [44]:
data.shape

(100000, 23)

In [71]:
# Label isolation
X = data.drop('Credit_Score', axis=1)
y = data['Credit_Score']

# Data splitting
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Model fitting
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(
    n_estimators=100,
    max_depth=5,
    random_state=42
)
model.fit(X_train, y_train)

# Model prediction
y_pred = model.predict(X_test)

# Model evaluation
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.58      0.39      0.47      3527
           1       0.65      0.65      0.65      5874
           2       0.66      0.74      0.70     10599

    accuracy                           0.65     20000
   macro avg       0.63      0.59      0.61     20000
weighted avg       0.65      0.65      0.64     20000

